Lib/test/test_logging.py

```diff
@@ -4310,6 +4312,31 @@ def test_issue_89047(self):
             s = f.format(r)
             self.assertNotIn('.1000', s)
 
+    def test_issue_102402_msecs_has_no_floating_point_precision_loss(self):
+        tests = (
+            # time_ns is approx. 2023-03-04 04:25:20 UTC
+            # (time_ns, expected_msecs_value)
+            (1_677_902_297_100_000_000, 100.0),  # exactly 100ms
+            (1_677_903_920_999_998_503, 999.0),  # check truncating doesn't round
+            (1_677_903_920_000_998_503, 0.0),  # check truncating doesn't round
+        )
+        for ns, want in tests:
+            with patch('time.time_ns') as patched_ns:
+                patched_ns.return_value = ns
+                record = logging.makeLogRecord({'msg': 'test'})
+            self.assertEqual(record.msecs, want)
+
+    def test_issue_102402_relativeCreated_has_higher_precision(self):
+        ns = 1_677_903_920_000_998_503  # approx. 2023-03-04 04:25:20 UTC
+        offset_ns = 200
+        with patch("time.time_ns") as patched_ns:
+            patched_ns.return_value = ns
+            importlib.reload(logging)  # force logging._startTime to use mocked value
```

Comment:
I am not sure that it does not have some hidden side effects. `logging` is not a single-file module, but a package, also other modules can keep references to objects from the old `logging`.

It would be better to import a fresh package (including submodules) and restore them after finishing the test. `test.support.import_helper` and `test.test_importlib.util` contain numerous helpers, but no one of them solves exactly this problem. The most reliable way is perhaps:
```
orig_modules = import_helper._save_and_remove_modules(['logging'])
try:
    import logging
    ... # test using logging
finally:
    import_helper._save_and_remove_modules(['logging'])
    sys.modules.update(orig_modules)
```
_Side note: It might be worth adding a less private helper for this. That's for another issue._

Reply:
Oooo, nice I didn't know `test.support.import_helper` existed. I'll have to read up on those.

Updated to your suggestion.:wq